### PR TITLE
Fix AST fuzzer issue in function `countMatches`

### LINKED ISSUE #59749: Logical error in `countMatches` found by AST Fuzzer
https://s3.amazonaws.com/clickhouse-test-reports/59563/a740fc7835ff54bda95c322d9d19e31349144d8e/ast_fuzzer__tsan_.html

```sql
SELECT  (39.08935546874999, 57.25528054528888), countMatches(toFixedString(materialize('foo'), 3), toFixedString(toFixedString('[f]{0}', 6), 6)) GROUP BY jumpConsistentHash(42, 57)
```

```
[...]
2024.02.07 18:10:03.407721 [ 1213 ] {} <Fatal> BaseDaemon: 12. DB::FunctionCountMatches<(anonymous namespace)::FunctionCountMatchesCaseSensitive>::executeImpl(std::vector<DB::ColumnWithTypeAndName, std::allocator<DB::ColumnWithTypeAndName>> const&, std::shared_ptr<DB::IDataType const> const&, unsigned long) const @ 0x00000000092fd4d6 in /workspace/clickhouse
2024.02.07 18:10:07.835696 [ 1213 ] {} <Fatal> BaseDaemon: 13. DB::IFunction::executeImplDryRun(std::vector<DB::ColumnWithTypeAndName, std::allocator<DB::ColumnWithTypeAndName>> const&, std::shared_ptr<DB::IDataType const> const&, unsigned long) const @ 0x0000000007453304 in /workspace/clickhouse
2024.02.07 18:10:12.281175 [ 1213 ] {} <Fatal> BaseDaemon: 14. DB::FunctionToExecutableFunctionAdaptor::executeDryRunImpl(std::vector<DB::ColumnWithTypeAndName, std::allocator<DB::ColumnWithTypeAndName>> const&, std::shared_ptr<DB::IDataType const> const&, unsigned long) const @ 0x0000000007452bb1 in /workspace/clickhouse
2024.02.07 18:10:12.324003 [ 1213 ] {} <Fatal> BaseDaemon: 15. ./build_docker/./src/Functions/IFunction.cpp:0: DB::IExecutableFunction::executeWithoutLowCardinalityColumns(std::vector<DB::ColumnWithTypeAndName, std::allocator<DB::ColumnWithTypeAndName>> const&, std::shared_ptr<DB::IDataType const> const&, unsigned long, bool) const @ 0x0000000016671fe7 in /workspace/clickhouse
[...]
```

### PR DESCRIPTION
**Pull request by rschu1ze**
Resolves #59749

Reviewer note: Contains
- Refactorings: https://github.com/ClickHouse/ClickHouse/commit/cfa1a452e1b3ea8b8f5c5c5af269ea0c0a35a8a9
- More refactorings and disallows `FixedString` patterns (which are obscure and disallowed elsewhere too): https://github.com/ClickHouse/ClickHouse/commit/f7bb000b5177a250bc44b3fadd6c5a33033ec0e2
- Fixed the issue which was caused by a missing implementation for `FixedString` matching: https://github.com/ClickHouse/ClickHouse/commit/fc003497bbea9b21294d77ad4e0f6b481240e07b

### Changelog category (leave one):
- Bug Fix (user-visible misbehavior in an official stable release)

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Fixed an exception in function `countMatches` with non-const `FixedString` haystack arguments, e.g. `SELECT countMatches(materialize(toFixedString('foobarfoo', 9)), 'foo');`

### CODE REVIEW

**Comment by robot-ch-test-poll** [comment]
*This is an automated comment for commit fc003497bbea9b21294d77ad4e0f6b481240e07b with description of existing statuses. It's updated for the latest CI running*

[⏳ Click here](https://s3.amazonaws.com/clickhouse-test-reports/59752/fc003497bbea9b21294d77ad4e0f6b481240e07b/ci_running.html) to open a full report in a separate page

<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>CI running</td><td>A meta-check that indicates the running CI. Normally, it's in <b>success</b> or <b>pending</b> state. The failed status indicates some problems with the PR</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/59752/fc003497bbea9b21294d77ad4e0f6b481240e07b/ci_running.html">⏳ pending</a></td></tr>
<tbody>
</table>

File: docs/en/sql-reference/functions/string-search-functions.md [hunk 1]
```diff
@@ -590,6 +590,10 @@ Result:
 └───────────────────────────────┘
 ```
 
+## countMatchesCaseInsensitive
```
Comment (jkartseva) [inline docs/en/sql-reference/functions/string-search-functions.md:593]: Should be added to `aspell-dict.txt` to whitelist (and unblock the tests).

File: tests/queries/0_stateless/01595_countMatches.sql [hunk 2]
```diff
@@ -25,5 +25,10 @@ select countMatchesCaseInsensitive('foo.com BAR.COM baz.com bam.com', '([^. ]+)\
...
 select 'errors';
-select countMatches(1, 'foo') from numbers(1); -- { serverError 43 }
-select countMatches('foobarfoo', toString(number)) from numbers(1); -- { serverError 44 }
+select countMatches(1, 'foo') from numbers(1); -- { serverError ILLEGAL_TYPE_OF_ARGUMENT }
+select countMatches('foobarfoo', toString(number)) from numbers(1); -- { serverError ILLEGAL_COLUMN }
+select countMatches('foo', materialize('foo')); -- { serverError ILLEGAL_COLUMN }
+
+select 'FixedString';
+select countMatches(toFixedString('foobarfoo', 9), 'foo');
+select countMatches(materialize(toFixedString('foobarfoo', 9)), 'foo');
```
Comment (jkartseva) [inline tests/queries/0_stateless/01595_countMatches.sql:34]: Please add tests for `String` and a constant, e.g.
``` sql
select countMatches(materialize(toString('fizzbuzz')), toString('fizz'));
select countMatches(materialize('fizzbuzz'), 'fizz');
```

File: src/Functions/countMatches.h [hunk 3]
```diff
@@ -83,13 +80,22 @@ class FunctionCountMatches : public IFunction
...
+            for (size_t i = 0; i < input_rows_count; ++i)
+            {
+                std::string_view str = col_haystack_fixedstring->getDataAt(i).toView();
+                vec_res[i] = countMatches(str, re, matches);
+            }
+
+            return col_res;
         }
         else
             throw Exception(ErrorCodes::LOGICAL_ERROR, "Error in FunctionCountMatches::getReturnTypeImpl()");
```
Comment (jkartseva) [inline src/Functions/countMatches.h:101]: This error message is not informative, it should indicate what went wrong, e.g. "Failed to cast haystack input to a string type (String, FixedString, a string literal)".

**Review by jkartseva — APPROVED** [review]

File: docs/en/sql-reference/functions/string-search-functions.md (hunk 1 above)
Comment (rschu1ze) [reply docs/en/sql-reference/functions/string-search-functions.md:593]: Done, thanks.

File: src/Functions/countMatches.h (hunk 3 above)
Comment (rschu1ze) [reply src/Functions/countMatches.h:101]: We should theoretically not even end up here (`LOGICAL_ERRROR`). But you are right, I improved the wording.

File: tests/queries/0_stateless/01595_countMatches.sql (hunk 2 above)
Comment (rschu1ze) [reply tests/queries/0_stateless/01595_countMatches.sql:36]: Done

**Comment by rschu1ze** [comment]
The remaining AST Fuzzer error ([here](https://s3.amazonaws.com/clickhouse-test-reports/59752/04f61781768b6f6bbd3a81e1c4e61d2d343c3410/ast_fuzzer__ubsan_.html)) complains that the pattern argument is non-const. This situation is actually impossible to exist since `getReturnTypeImpl()` rejects non-const pattern arguments. I tried with the causing query according to the log in an ubsan build

```sql
SELECT countMatches(materialize('foo.com bar.com baz.com bam.com'), '([^. ]+)\\.([^. ]+)') AS `countMatches(materialize('foo.com bar.com baz.com bam.com'), '([^. ]+)\\\\.([^. ]+)')` FROM system.one AS __table1 GROUP BY '([^. ]+)\\.([^. ]+)', 'foo.com bar.com baz.com bam.com', '([^. ]+)\\.([^. ]+)'
```

but the issue did not reproduce. Anyways, I added another check, let's see ...

**Comment by rschu1ze** [comment]
> Anyways, I added another check, let's see ...

This helped only in the sense that we are no longer killed by the sanitizer, yet the ast fuzzer ([here](https://s3.amazonaws.com/clickhouse-test-reports/59752/b1099e181d178072daffcf78d9dfa2fe8a97cd73/ast_fuzzer__asan_.html)) still fails due to an exception. Like before, the problem did not reproduce for me locally using the failing query according to server logs `SELECT countMatches(materialize(toLowCardinality(toFixedString('foo', 3))), '[f]{0}foo') GROUP BY '[f]{0}foo', toFixedString(toFixedString(toFixedString(toFixedString(toFixedString(toFixedString(toFixedString(toNullable('foobarfoo'), 9), 9), toNullable(9)), 9), 9), toLowCardinality(9)), toLowCardinality(9)), toLowCardinality('foo'), materialize('foobarfoo')`. And also like before, we should not end up with an exception as this code running before rejects non-const patterns:

```cpp
DataTypePtr getReturnTypeImpl(const ColumnsWithTypeAndName & arguments) const override
{
    FunctionArgumentDescriptors args{
        {"haystack", &isStringOrFixedString<IDataType>, nullptr, "String or FixedString"},
        {"pattern", &isString<IDataType>, isColumnConst, "constant String"}
    };
    validateFunctionArgumentTypes(*this, arguments, args);

    return std::make_shared<DataTypeUInt64>();
}
```

To me, this looks like an issue with the AST fuzzer itself, @Algunenano I vaguely remember you improved it recently, perhaps you like to take a look?

**Comment by Algunenano** [comment]
> To me, this looks like an issue with the AST fuzzer itself, @Algunenano I vaguely remember you improved it recently, perhaps you like to take a look?

Not related to the fuzzer at all. The fuzzer just generates queries. It's the server that's detecting the assert and crashing.


>  Like before, the problem did not reproduce for me locally using the failing query according to server logs

Your build doesn't have asserts enabled. Debug build has them on by default:

```
$ ./build_debug/programs/clickhouse local
ClickHouse local version 24.2.1.1.

Mordor :) SELECT countMatches(materialize(toLowCardinality(toFixedString('foo', 3))), '[f]{0}foo') GROUP BY '[f]{0}foo', toFixedString(toFixedString(toFixedString(toFixedString(toFixedString(toFixedString(toFixedString(toNullable('foobarfoo'), 9), 9), toNullable(9)), 9), 9), toLowCardinality(9)), toLowCardinality(9)), toLowCardinality('foo'), materialize('foobarfoo')

SELECT countMatches(materialize(toLowCardinality(toFixedString('foo', 3))), '[f]{0}foo')
GROUP BY
    '[f]{0}foo',
    toFixedString(toFixedString(toFixedString(toFixedString(toFixedString(toFixedString(toFixedString(toNullable('foobarfoo'), 9), 9), toNullable(9)), 9), 9), toLowCardinality(9)), toLowCardinality(9)),
    toLowCardinality('foo'),
    materialize('foobarfoo')

Query id: cdc9e600-25c6-4101-8f82-3904662c428b

2024.02.12 11:40:02.258234 [ 19572 ] {cdc9e600-25c6-4101-8f82-3904662c428b} <Fatal> : Logical error: 'Error in FunctionCountMatches::getReturnTypeImpl()'.
Aborted (core dumped)
```

**Comment by alexey-milovidov** [comment]
> Logical error: 'Pattern argument is not const'.

**Comment by rschu1ze** [comment]
Thanks. I still could not reproduce the issue locally despite having a debug build (i.e. no exception or crash), but looking at the server log again shows that `getReturnTypeImpl()` is not exercised at all. I can dig deeper why that is but for now, I changed the exception kind which should prevent issues in CI.

**Comment by rschu1ze** [comment]
ClickHouse AST Fuzzer (ubsan): #58979
ClickHouse Stress Test (debug): Looks like #59894 but the issue did not reproduce with the failing query locally for me.

**Pull request merged**